File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int st = 0, ans = 0, maxFreq = 0;
        int n = s.length();
        Map<Character, Integer> mpp = new HashMap<>();

        for (int i = 0; i < n; i++) {
            mpp.put(s.charAt(i), mpp.getOrDefault(s.charAt(i), 0) + 1);
            maxFreq = Math.max(maxFreq, mpp.get(s.charAt(i)));

            while (i - st + 1 - maxFreq > k) {
                mpp.put(s.charAt(st), mpp.get(s.charAt(st)) - 1);

                if (mpp.get(s.charAt(st)) == 0) {
                    mpp.remove(s.charAt(st));
                }

                st++;
            }

            ans = Math.max(ans, i - st + 1);
        }

        return ans;
    }
}

File: 424-longest-repeating-character-replacement/longest-repeating-character-replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {
        int st = 0;
        int ans = 0;
        map<char, int> mpp;
        int n = s.size();
        int maxFreq = 0;

        for(int i = 0; i < n; i++){
            mpp[s[i]]++;
            maxFreq = max(mpp[s[i]], maxFreq);

            while(i - st + 1 - maxFreq > k){
                mpp[s[st]]--;

                if(mpp[s[i]] == 0){
                    mpp.erase(mpp[s[i]]);
                }

                st++;
            }

            ans = max(ans, i - st + 1);

        }

        return ans;
    }
};